### DISCUSSION TITLE
Snapshot: only running 3 devices at a time?

### DISCUSSION THREAD

**Discussion by benomatis** [fastlane/fastlane#18582] [Q&A]
I first thought snapshot is just ignoring the last one of a 4 devices list in my Snapfile, but then I added a 5th one as a workaround, thinking it's only ignoring the last one, but it turns out it will only launch 3 simulators at a time, however long my devices list is. Is this a setting somewhere or will it do it in a batch of 3 and return to the rest once the first 3 is done...?

**Comment by rodrigoelp** [comment]
Hey, I came to ask exactly that question.

I read in the documentation that fastlane it is meant to launch as many simulators as cores available but, I got access to a macpro with 16 cores (32 threads) and it never gets to launch more than 3 simulators during the snapshot process.

Then I thought I could bypass this issue by creating a test plan with parallel test execution and splitting my app into multiple chunks to take screenshots on parallel devices but fastlane seems to ignore the configuration of parallel execution (which is ok I guess).
Checking the [test command generator base](https://github.com/fastlane/fastlane/blob/7cc8c67a42ada48d83f1d1c25d3f2a8c89d0c3c2/snapshot/lib/snapshot/test_command_generator_base.rb) and its [implementation](https://github.com/fastlane/fastlane/blob/7cc8c67a42ada48d83f1d1c25d3f2a8c89d0c3c2/snapshot/lib/snapshot/test_command_generator.rb) I could not find anywhere the specification of the devices but I can see that, regardless of the number of devices I tell it, fastlane issues the command as follows:

```
set -o pipefail && xcodebuild -scheme SampleScreenshots -project ./SampleScreenshots.xcodeproj -derivedDataPath ./output/derived_data -maximum-concurrent-test-simulator-destinations 8 -destination 'platform=iOS Simulator,name=iPhone 11,OS=14.5' -destination 'platform=iOS Simulator,name=iPhone 11 Pro,OS=14.5' -destination 'platform=iOS Simulator,name=iPhone 12,OS=14.5' FASTLANE_SNAPSHOT=YES FASTLANE_LANGUAGE=en-US build test | tee ../snapshot/SampleScreenshots-SampleScreenshots.log | xcpretty
```
Followed by a 

```
set -o pipefail && xcodebuild -scheme SampleScreenshots -project ./SampleScreenshots.xcodeproj -derivedDataPath ./output/derived_data -maximum-concurrent-test-simulator-destinations 8 -destination 'platform=iOS Simulator,name=iPhone 12 mini,OS=14.5' -destination 'platform=iOS Simulator,name=iPhone 8,OS=14.5' FASTLANE_SNAPSHOT=YES FASTLANE_LANGUAGE=en-US build test | tee -a ../snapshot/SampleScreenshots-SampleScreenshots.log | xcpretty
```

I have been reading fastlane for a week now trying to fix the issue myself :|

**Reply by rodrigoelp** [reply]
Providing some update, finally found where the grouping happens: https://github.com/fastlane/fastlane/blob/7cc8c67a42ada48d83f1d1c25d3f2a8c89d0c3c2/snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb#L63

In there you will see they try to launch as many cores - 1 to run the tests.

Fastlane is looking at the physical cores as opposed to the logical cores (which is fine).
The configuration I was running on my old mac pro fails for some but I did manage to get it to work as expected on one of the sample projects, launching 5 simulators at the same time.
